tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -840,10 +834,44 @@ private[optimizer] abstract class OptimizerCore(
       case tree: BinaryOp =>
         pretransformBinaryOp(tree)(cont)
 
+      case tree: JSBracketSelect =>
+        pretransformJSBracketSelect(tree)(cont)
+
       case tree: JSFunctionApply =>
         pretransformJSFunctionApply(tree, isStat = false,
             usePreTransform = true)(cont)
 
+      case JSArrayConstr(items) =>
+        if (items.exists(_.isInstanceOf[JSSpread])) {
+          /* TODO This means spread in array constr does not compose under
+           * this optimization. We could improve this with a
+           * pretransformExprsOrSpreads() or something like that.
+           */
+          cont(JSArrayConstr(transformExprsOrSpreads(items)).toPreTransform)
+        } else {
+          pretransformExprs(items) { titems =>
+            tryOrRollback { cancelFun =>
```

Comment:
In a codebase "closer to JS", we wouldn't get here in `pretransformExpr(JSArrayConstr)`. The `JSArrayConstr` would probably be a direct argument to a JS method call, in which case it would be transformed by `transformExpr(JSArrayConstr)`, which does not try to optimize it with a `tryOrRollback`.

It would still happen if the developer stores the `js.Array()` in a `val` first, though.
```scala
val a = js.Array(1, 2)
jsObj.foo(a)
```
would have a rollback, indeed. But
```scala
jsObj.foo(js.Array(1, 2))
```
wouldn't.